File: RESTFul/ADS/Java/app1/src/main/java/com/att/api/ads/service/ADSService.java
/*
 * Copyright 2015 AT&T
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.att.api.ads.service;

import com.att.api.oauth.OAuthToken;
import com.att.api.rest.APIResponse;
import com.att.api.rest.RESTClient;
import com.att.api.rest.RESTException;
import com.att.api.service.APIService;

import org.json.JSONObject;
import org.json.JSONException;


/**
 * Used to interact with version 1 of the Advertising API.
 *
 * <p>
 * This class is thread safe.
 * </p>
 *
 * @author pk9069
 * @version 1.0
 * @since 1.0
 * @see <a href="https://developer.att.com/docs/apis/rest/1/Advertising">Advertising Documentation</a>
 */
public class ADSService extends APIService {

    /**
     * Converts an array of keywords to a keywords string to be sent to the
     * API.
     *
     * @param keywords keywords
     * 
     * @return keywords as a string
     */
    private String getKeywordsStr(final String[] keywords) {
        if (keywords == null) {
            return null;
        }

        final StringBuilder sb = new StringBuilder();
        final int length = keywords.length;
        for (int i = 0; i < length - 1; ++i) {
            sb.append(keywords[i] + ",");
        }
        sb.append(keywords[length - 1]);

        return sb.toString();
    }

    /**
     * Appends any optional arguments to the request.
     *
     * @param args optional arguments to append
     * @param client rest client to append arguments to
     */
    private void appendArguments(ADSArguments args, RESTClient client) {
        // A lot of this could probably be moved to array...

        if (args.getGender() != null)
            client.addParameter("Gender", args.getGender().getString());

        if (args.getZipCode() != -1)
            client.addParameter("ZipCode", "" + args.getZipCode());

        if (args.getAreaCode() != -1)
            client.addParameter("AreaCode", "" + args.getAreaCode());

        if (args.getCity() != null)
            client.addParameter("City", args.getCity());

        if (args.getCountry() != null)
            client.addParameter("Country", args.getCountry());

        if (args.getLongitude() != -1)
            client.addParameter("Longitude", "" + args.getLongitude());

        if (args.getLatitude() != -1)
            client.addParameter("Latitude", "" + args.getLatitude());

        if (args.getType() != null)
            client.addParameter("Type", "" + args.getType().getValue());

        if (args.getAgeGroup() != null)
            client.addParameter("AgeGroup", args.getAgeGroup());

        if (args.getOver18() != null)
            client.addParameter("Over18", "" + args.getOver18().getValue());

        final String keywordsStr = getKeywordsStr(args.getKeywords());
        if (keywordsStr != null)
            client.addParameter("Keywords", keywordsStr);

        if (args.getPremium() != null)
            client.addParameter("Premium", "" + args.getPremium().getValue());

        if (args.getMaxHeight() != -1)
            client.addParameter("MaxHeight", "" + args.getMaxHeight());

        if (args.getMaxWidth() != -1)
            client.addParameter("MaxWidth", "" + args.getMaxWidth());

        if (args.getMinHeight() != -1)
            client.addParameter("MinHeight", "" + args.getMinHeight());

        if (args.getMinWidth() != -1)
            client.addParameter("MinWidth", "" + args.getMinWidth());

    }

    /**
     * Creates an ADService object.
     *
     * @param fqdn fully qualified domain name to use for sending requests
     * @param token OAuth token to use for authorization
     */
    public ADSService(String fqdn, OAuthToken token) {
        super(fqdn, token);
    }

    /**
     * Convenience method for sending a request to the API for getting an
     * advertisement with only the required values set.
     *
     * @param category category of advertisement
     * @param userAgent user agent
     * @param udid universally unique identifier
     * @return API response
     * @throws RESTException if API request was not successful
     * @throws IllegalArgumentException if args is null
     */
    public ADSResponse getAdvertisement(Category category, String userAgent,
            String udid) throws RESTException {

        final ADSArguments args 
            = new ADSArguments.Builder(category, userAgent, udid).build();
        
        return this.getAdvertisement(args);
    }

    /**
     * Sends a request to the API for getting an advertisement.
     *
     * @param args arguments 
     * @return API response
     * @throws RESTException if API request was not successful
     * @throws IllegalArgumentException if args is null
     */
    public ADSResponse getAdvertisement(ADSArguments args) throws RESTException {

        if (args == null)
            throw new IllegalArgumentException("Arguments must not be null.");

        final String endpoint = getFQDN() + "/rest/1/ads";

        final RESTClient client = new RESTClient(endpoint)
            .addParameter("Category", args.getCategory().getString())
            .addAuthorizationHeader(getToken())
            .addHeader("User-Agent", args.getUserAgent())
            .addHeader("UDID", args.getUdid());

        this.appendArguments(args, client);

        try {
            final APIResponse response = client.httpGet();
            if (response.getStatusCode() == 204) {
                return null;
            }
            final String responseBody = response.getResponseBody();
            return ADSResponse.valueOf(new JSONObject(responseBody));
        } catch (JSONException pe) {
            throw new RESTException(pe);
        }
    }
}

/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4 foldmethod=marker */

File: RESTFul/ADS/PHP/app1/lib/ADS/ADSService.php
<?php
namespace Att\Api\ADS;

/*
 * Copyright 2014 AT&T
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

require_once __DIR__ . '../../Srvc/APIService.php';
require_once __DIR__ . '/ADSResponse.php';

use Att\Api\Restful\RestfulRequest;
use Att\Api\Restful\HttpGet;
use Att\Api\Srvc\APIService;
use Att\Api\Srvc\Service;
use Att\Api\OAuth\OAuthToken;

/**
 * Used to interact with version 1 of the Advertising API.
 *
 * For a list of acceptable values and their definitions, refer to 
 * {@link https://developer.att.com/docs/apis/rest/1/Advertising}.
 *
 * @category API
 * @package  ADS
 * @author   pk9069
 * @license  http://www.apache.org/licenses/LICENSE-2.0
 * @version  Release: @package_version@ 
 * @link     https://developer.att.com/docs/apis/rest/1/Advertising
 */
class ADSService extends APIService
{

    /**
     * Convenience method to append any optional arguments to the specified
     * <var>$httpGet</var> object.
     *
     * @param HttpGet $httpGet HttpGet object to append to
     * @param OptArgs $optArgs optional arguments used for appending
     *
     * @return void
     */
    private function _appendOptArgs(HttpGet $httpGet, OptArgs $optArgs)
    {
        $keys = array(
            'AgeGroup', 'AreaCode', 'City', 'Country', 'Gender',
            'Keywords', 'Latitude', 'Longitude', 'MaxHeight', 'MaxWidth',
            'MinHeight', 'MinWidth', 'Type', 'ZipCode'
        );

        $keywords = $optArgs->getKeywords();
        if ($keywords !== null)
            $keywords = implode(',', $optArgs->getKeywords());

        $vals = array(
            $optArgs->getAgeGroup(), $optArgs->getAreaCode(),
            $optArgs->getCity(), $optArgs->getCountry(), $optArgs->getGender(),
            $keywords, $optArgs->getLatitude(), $optArgs->getLongitude(),
            $optArgs->getMaxHeight(), $optArgs->getMaxWidth(),
            $optArgs->getMinHeight(), $optArgs->getMinWidth(),
            $optArgs->getAdType(), $optArgs->getZipCode()
        );

        for ($i = 0; $i < count($keys); ++$i) {
            $key = $keys[$i];
            $val = $vals[$i];

            if ($val != null) {
                $httpGet->setParam($key, $val);
            }
        }
    }

    /**
     * Creates an ADSService object that can be used to interact with
     * the ADS API.
     *
     * @param string     $FQDN  fully qualified domain name to which requests 
     *                          will be sent
     * @param OAuthToken $token OAuth token used for authorization 
     */
    public function __construct($FQDN, OAuthToken $token)
    {
        parent::__construct($FQDN, $token); 
    }

    /**
     * Sends a request to the API for getting an advertisement. 
     * 
     * @param string       $category  category of this app.
     * @param string       $userAgent user agent string to send to API.
     * @param string       $udid      specifies a universially unique
     *                                identifier, which must be at least 30
     *                                characters in length.
     * @param OptArgs|null $optArgs   any optional values.
     *
     * @return null|ADSResponse null if no ads were returned, 
     *                          otherwise an ADSResponse object
     * @throws ServiceException if API request was not successful
     */
    public function getAdvertisement(
        $category, $userAgent, $udid, OptArgs $optArgs = null
    ) {

        $endpoint = $this->getFqdn() . '/rest/1/ads';

        $req = new RestfulRequest($endpoint);
        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('User-agent', $userAgent)
            ->setHeader('Udid', $udid);

        $httpGet = new HttpGet();
        $httpGet->setParam('Category', $category);

        if ($optArgs != null)
            $this->_appendOptArgs($httpGet, $optArgs);

        $result = $req->sendHttpGet($httpGet);

        // no ads returned
        if ($result->getResponseCode() == 204) {
            return null;
        }

        // response as json array
        $jarr = Service::parseJson($result);

        return ADSResponse::fromArray($jarr);
    }

}

/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4: */
?>
